### ISSUE TITLE
extensions/v1beta1 Ingress in Rancher chart is deprecated

### ISSUE THREAD

**Issue by jiaqiluo** [rancher/rancher#29144]
**What kind of request is this (question/bug/enhancement/feature request):**
enhancement

**Steps to reproduce (least amount of steps as possible):**
- install Rancher: `v2.5.0-rc1` in airgap env in HA mode 
- here is the helm command to generate the template
```
 helm template rancher ./rancher-2.5.0-rc1.tgz --output-dir . \
    --namespace cattle-system \
    --set hostname=jiaqi0235-15555-internal.qa.rancher.space \
    --set rancherImage=ec2-18-222-107-174.us-east-2.compute.amazonaws.com/rancher/rancher \
    --set ingress.tls.source=secret \
    --set systemDefaultRegistry=ec2-18-222-107-174.us-east-2.compute.amazonaws.com \
    --set useBundledSystemChart=true
```
- and install rancher 
```
kubectl create namespace cattle-system
kubectl -n cattle-system apply -R -f ./rancher
```

**Result:**
```
>  kubectl -n cattle-system apply -R -f ./rancher
clusterrolebinding.rbac.authorization.k8s.io/rancher created
deployment.apps/rancher created
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
ingress.extensions/rancher created
service/rancher created
serviceaccount/rancher created
```

we should update the Rancher helm chart to use `networking.k8s.io/v1` for Ingress

**Comment by brendarearden** [comment]
is available in the most recent master and 2.5 image tags for testing

**Comment by jiaqiluo** [comment]
wait for a v2.5.2-rc to validate this fix in the helm chart.

**Comment by mrajashree** [comment]
Can be tested with v2.5.2-rc1

**Comment by jiaqiluo** [comment]
in `v2.5.2-rc1` HA air gap installation 

the abovementioned warning for `extensions/v1beta1` is gone, but the following new warning appears

> Warning: networking.k8s.io/v1beta1 Ingress is deprecated in v1.19+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress

and the recommended version `networking.k8s.io/v1` is the same 

the full output:
```
>  kubectl -n cattle-system apply -R -f ./rancher
clusterrolebinding.rbac.authorization.k8s.io/rancher unchanged
deployment.apps/rancher created
Warning: networking.k8s.io/v1beta1 Ingress is deprecated in v1.19+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
ingress.networking.k8s.io/rancher created
issuer.cert-manager.io/rancher created
service/rancher created
serviceaccount/rancher created
```

**Comment by sowmyav27** [comment]
PR: https://github.com/rancher/rancher/pull/29878

**Comment by sowmyav27** [comment]
Reopening this issue Because of issue seen - https://github.com/rancher/rancher/issues/29874

**Comment by izaac** [comment]
Rancher version: `v2.5.2-rc5 (11/01/2020)`
Helm version: `v3.4.0`
cert-manager version: `v1.0.4`

Reopening, I'm getting this error. I was following the docs with a Rancher generated certificates option:
https://rancher.com/docs/rancher/v2.x/en/installation/other-installation-methods/air-gap/install-rancher/




```
helm template cert-manager ./cert-manager-v1.0.4.tgz --output-dir . \
  --namespace cert-manager \
  --set image.repository=<private_registry_url>/quay.io/jetstack/cert-manager-controller \
  --set webhook.image.repository=<private_registry_url>/quay.io/jetstack/cert-manager-webhook \
  --set cainjector.image.repository=<private_registry_url>/quay.io/jetstack/cert-manager-cainjector
```

cert-manager CRDs are installed from here: https://github.com/jetstack/cert-manager/releases/download/v1.0.4/cert-manager.crds.yaml

```
curl -L -o cert-manager/cert-manager-crd.yaml https://github.com/jetstack/cert-manager/releases/download/v1.0.4/cert-manager.crds.yaml
```

Install cert-manager

```
kubectl create namespace cert-manager
kubectl apply -R -f ./cert-manager
(wait for pods to be active - kubectl get po -n cert-manager)
```



```
helm template rancher ./rancher-2.5.2-rc5.tgz --output-dir . \
  --namespace cattle-system \
  --set hostname=<rancher_host> \
  --set certmanager.version=v1.0.4 \
  --set rancherImage=<private_registry_url>/rancher/rancher \
  --set systemDefaultRegistry=<private_registry_url> \
  --set useBundledSystemChart=true
```

Install Rancher
```
kubectl create namespace cattle-system
kubectl -n cattle-system apply -R -f ./rancher
clusterrolebinding.rbac.authorization.k8s.io/rancher created
deployment.apps/rancher created
Warning: networking.k8s.io/v1beta1 Ingress is deprecated in v1.19+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
ingress.networking.k8s.io/rancher created
service/rancher created
serviceaccount/rancher created
error: unable to recognize "rancher/templates/issuer-rancher.yaml": no matches for kind "Issuer" in version "certmanager.k8s.io/v1alpha1"
```

`cat rancher/templates/issuer-rancher.yaml`
```
---
# Source: rancher/templates/issuer-rancher.yaml
apiVersion: certmanager.k8s.io/v1alpha1
kind: Issuer
metadata:
  name: rancher
  labels:
    app: rancher
    chart: rancher-2.5.2-rc5
    heritage: Helm
    release: rancher
spec:
  ca:
    secretName: tls-rancher
```

**Comment by izaac** [comment]
kubectl version

```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-20T14:12:21Z", GoVersion:"go1.15.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-14T12:41:49Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by kinarashah** [comment]
Available to test in `v2.5.2-rc6`

**Comment by izaac** [comment]
After checking the logic on this code https://github.com/rancher/rancher/blob/release/v2.5/chart/templates/issuer-rancher.yaml I realized I was using the wrong `--set certmanager.version` parameter value
Before:
`--set certmanager.version=v1.0.1`
After:
`--set certmanager.version=1.0.1`
12:55
Rancher `v2.5.2-rc6` is working now.

**Comment by brendarearden** [comment]
After looking at this some more, it seems if we revert this [revert](https://github.com/rancher/rancher/pull/29888), then it would work on a `helm install`. It appears to be an issue with `.Capabilities.APIVersions.Has` when running `helm template` based on this [comment](https://github.com/roboll/helmfile/issues/1014#issuecomment-561201038)

**Comment by brendarearden** [comment]
@izaac if you add `-a networking.k8s.io/v1/Ingress` to the `helm template` command, it provides the api-version to be evaluated for the built in helm object. https://helm.sh/docs/helm/helm_template/ essentially you have to tell helm what api-versions are available to it if you want it to evaluate everything properly. 

After we get the revert merged, then try this: 
```
helm template rancher ./rancher-2.5.2-rc5.tgz --output-dir . \
  --namespace cattle-system \
  --api-versions networking.k8s.io/v1/Ingress \
  --set hostname=<rancher_host> \
  --set certmanager.version=v1.0.4 \
  --set rancherImage=<private_registry_url>/rancher/rancher \
  --set systemDefaultRegistry=<private_registry_url> \
  --set useBundledSystemChart=true
```

**Comment by izaac** [comment]
@brendarearden ok that is new for this Rancher installation method as it's not documented here: https://rancher.com/docs/rancher/v2.x/en/installation/other-installation-methods/air-gap/install-rancher/

But I will try it once it's available to test again.

**Comment by cbron** [comment]
We don't want to add it to the official instructions because it's basically faking the available API endpoints. I'm surprised we use `helm template` and not `helm install`, as `helm install` does those checks for us.

**Comment by brendarearden** [comment]
Changes have been added back in. Since 1.19 is the only version out of the currently support k8s versions that serves the ingress resource from v1, I left the default as `networking.k8s.io/v1beta1` so the warning will still show up on 1.19 install unless you pass in the api-version. We will not be updating the official documentation for `helm template` - if a user knows their apiVersion when running help template, and they want to get rid of the warning on a 1.19 install they could use the `--api-versions networking.k8s.io/v1/Ingress`.

**Comment by cbron** [comment]
Looking at this again I think the warning will be removed as it is `extensions/v1beta1` that was causing the warning, _not_ `networking.k8s.io/v1beta1`. That one is not deprecated as far as I can tell.

**Comment by jiaqiluo** [comment]
@cbron  That is right, the warning shows only when `extensions/v1beta1` is used for the ingress in Rancher's chart.  Just like @brendarearden suggests, we can pass the `--api-versions` flag to set the API version to avoid the warning. 

As to why we use `helm template` not `helm install`, we were using Helm 2 when the documentation was written, so we had to render the template because there is no tiller running on the local cluster. 

Helm 3 supports installing the local packaged chart, which we have not validated yet. I created an issue [here](https://github.com/rancher/rancher/issues/30114) for validating the `helm install` approach.

**Comment by jiaqiluo** [comment]
Waiting for an RC.

**Comment by moophlo** [comment]
How can i upgrade my existing Ingresses in the Default project to use the new api?

kubectl get ingress wordpress
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
NAME        CLASS    HOSTS              ADDRESS                                                             PORTS     AGE
wordpress   <none>   www.mufloland.it   docker01.mufloland.it,docker02.mufloland.it,docker03.mufloland.it   80, 443   99d

EDIT DOESN'T WORK:

# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    cert-manager.io/cluster-issuer: letsencrypt
    field.cattle.io/ingressState: '{"d29yZHByZXNzL2RlZmF1bHQvd3d3Lm11ZmxvbGFuZC5pdC8vLzgwODA=":"deployment:default:wordpress"}'
    field.cattle.io/publicEndpoints: '[{"addresses":["","",""],"port":443,"protocol":"HTTPS","serviceName":"default:ingress-23f9ea52b31dc3b5e25a8b7eb2b8ea14","ingressName":"default:wordpress","hostname":"www.mufloland.it","path":"/","allNodes":false}]'
    kubernetes.io/tls-acme: '"true"'
  creationTimestamp: "2020-08-10T12:10:11Z"
  generation: 10
  labels:
    app.kubernetes.io/instance: wordpress
    app.kubernetes.io/managed-by: Tiller
    app.kubernetes.io/name: wordpress
    helm.sh/chart: wordpress-9.0.3
    io.cattle.field/appId: wordpress
  name: wordpress
  namespace: default
  resourceVersion: "76171175"
  selfLink: /apis/extensions/v1beta1/namespaces/default/ingresses/wordpress
  uid: e909789f-417c-4be5-b776-58bfd5609f40
spec:
  rules:
  - host: www.mufloland.it
    http:
      paths:
      - backend:
          serviceName: ingress-23f9ea52b31dc3b5e25a8b7eb2b8ea14
          servicePort: 8080
        path: /
        pathType: ImplementationSpecific
  tls:
  - hosts:
    - www.mufloland.it
    secretName: wwwmuflolandit
status:
  loadBalancer:
    ingress:
    - hostname: docker01.mufloland.it
    - hostname: docker02.mufloland.it
    - hostname: docker03.mufloland.it
~

**Comment by moophlo** [comment]
@cbron how? how can i change it?

Take a look:

muflo@diapolinux ~/Scaricati/BNL/mybank $ kubectl get ingresses.v1.networking.k8s.io
NAME        CLASS    HOSTS                   ADDRESS                                                             PORTS     AGE
hello       <none>   hello.mufloland.it      docker01.mufloland.it,docker02.mufloland.it,docker03.mufloland.it   80, 443   6h56m
registry    <none>   registry.mufloland.it   docker01.mufloland.it,docker02.mufloland.it,docker03.mufloland.it   80, 443   99d
wordpress   <none>   www.mufloland.it        docker01.mufloland.it,docker02.mufloland.it,docker03.mufloland.it   80, 443   100d
muflo@diapolinux ~/Scaricati/BNL/mybank $ kubectl get ingresses.v1beta1.extensions
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
NAME        CLASS    HOSTS                   ADDRESS                                                             PORTS     AGE
hello       <none>   hello.mufloland.it      docker01.mufloland.it,docker02.mufloland.it,docker03.mufloland.it   80, 443   6h56m
registry    <none>   registry.mufloland.it   docker01.mufloland.it,docker02.mufloland.it,docker03.mufloland.it   80, 443   99d
wordpress   <none>   www.mufloland.it        docker01.mufloland.it,docker02.mufloland.it,docker03.mufloland.it   80, 443   100d
muflo@diapolinux ~/Scaricati/BNL/mybank $

**Comment by cbron** [comment]
@moophlo wordpress was probably installed via a chart, so you'd need to change the chart and then upgrade with helm

**Comment by jiaqiluo** [comment]
The fix is validated on rancher:`v2.5.4-rc2`.

With the fix, the ingress will be created with either the apiVersion `networking.k8s.io/v1` or `networking.k8s.io/v1beta1` based on the capability of the cluster ref: (https://github.com/brendarearden/rancher/blob/fd980b0d6eb6f2f450c06fe67e26e5e9186a4975/chart/templates/ingress.yaml#L1-L4)

But for the backward compatibility reason, until `extensions` is removed, `kubectl get ingress` uses `extensions/v1beta1` endpoint by default. ref: https://github.com/kubernetes/kubernetes/issues/62283#issuecomment-380968868

As the result, we may still see the warning about deprecated apiVersion in the output, but the ingress is actually created with the new apiVersion, and it can be checked by the following command
```
kubectl get ingress.networking -n cattle-system
```

### When using the `helm install` command to install Rancher 

We can check the same ingress with different API versions 


notice that there is a warning message in one of  the following output:
```
>  k get ingress.networking -n cattle-system rancher -o yaml | head -3
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:

> k get ingress -n cattle-system rancher -o yaml | head -3
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
```

### When using `helm` to generate the template then using `kubectl apply` to install Rancher 

In the generated template file, the ingress uses the `apiVersion` `networking.k8s.io/v1beta1`, which matches the logic in the design.
```
>  cat rancher/templates/ingress.yaml | head -3
---
# Source: rancher/templates/ingress.yaml
apiVersion: networking.k8s.io/v1beta1
```

when applying files, we see a warning message 
```
>  k apply -f rancher/templates/
clusterrolebinding.rbac.authorization.k8s.io/rancher created
deployment.apps/rancher created
Warning: networking.k8s.io/v1beta1 Ingress is deprecated in v1.19+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
ingress.networking.k8s.io/rancher created
issuer.cert-manager.io/rancher created
service/rancher created
serviceaccount/rancher created
```

and we can also check the ingress object with two api versions 
```
>  k get ingress.networking -n cattle-system rancher -o yaml | head -3
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:

> k get ingress -n cattle-system rancher -o yaml | head -3
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
```
